src/3pc/directory/impl.hpp

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2021 MIT Digital Currency Initiative,
+//                    Federal Reserve Bank of Boston
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef OPENCBDC_TX_SRC_3PC_DIRECTORY_IMPL_H_
+#define OPENCBDC_TX_SRC_3PC_DIRECTORY_IMPL_H_
+
+#include "interface.hpp"
+
+namespace cbdc::threepc::directory {
+    /// Implementation of a directory. Uses siphash to uniformly map keys to
```

Comment:
`... uniformly map keys to shard IDs based on a fixed number of shards`: seems correct to me (perhaps out of ignorance). @HalosGhost: if a value-adding correction is needed please make the change.

Reply:
Sorry, allow me to clarify: [`impl.cpp`:15](https://github.com/mit-dci/opencbdc-tx/pull/213/files#diff-77a7dfffad078695ca8ec3a68263414d82b8790b84ffdc33b307689e2b69eba9R15) divides the key generated by the number of shards running to select the shard for load-balancing. This constrains the random-number-generator of the hash function to the range of running shard ids (i.e., `[0, m_n_shards - 1]`). However, it introduces "modulo bias".[^1]

[^1]: https://stackoverflow.com/questions/10984974/why-do-people-say-there-is-modulo-bias-when-using-a-random-number-generator

I mentioned this because the comment is inaccurate. However, as I mentioned, I do not know if any security properties of the system depend on this property (i.e., I am not sure if this load-balancing is required to be perfectly uniformly distributed). If it does not matter, and this is either acceptably uniform for the purposes of the architecture, or the uniformity of the distribution is only desirable, not necessarily something that must be achieved, then that's not a problem! The comment should be corrected though.

However, if the load-balancing here is needed to be uniformly random, I see two valid directions: 1) we should document that the number of shards is required to be a multiple of the range of the hash-function. This is equivalent to taking the `n`-least-significant bits as the shard index. Or, 2) we modify `key_location()` cut the bias from this selection. This could be done either with the STL's templates for random-number generation (which we use extensively elsewhere in this codebase, for easy reference), or, say with rejection sampling into a large range known to be a multiple of `m_n_shards`, and then taking the modulus just as it does now. That also should perfectly remove the bias since it creates an equivalent case to `m_n_shards` being a multiple of the hash-function's range.

If we don't know, or don't need to know at this level of depth, whether or not this property is important to the system's security properties, we could potentially close this task by modifying the comment to be accurate. That's not a hard fix or unreasonable, but I don't know if it's the right choice. Definitely happy for alternative actions too, of course.